### PR TITLE
DMT-54 Add semi circle functionality

### LINKED ISSUE #49: Semi circle donut chart
Would be nice to be able to change donut chart to a semi circle. Semi Circle Donut Chart is simply a Donut Chart cut in half. The elements are still divided into sectors, illustrating numerical proportion like a basic pie or donut chart.

DVP_1_100-76

### PR DESCRIPTION
**Pull request by eemilj**
## Related issue
- Closes #49

## Proposed changes
- Add semi-circle option to popout settings
- Add functionality to visualize semi-circle and correspondingly adjust the location to center it
- Update logic for center-text to properly adjust according to circle type
- Refactor center-text into it's own class
- Fix labels offsets for semi-circle when placed outside circle

### CODE REVIEW

File: src/renderer.js [hunk 1]
```diff
@@ -146,90 +123,59 @@ export async function render(donutState, modProperty) {
...
+        return null;
+    }
+
     // If editing mode is enabled initialize the setting-popout
     donutState.context.isEditing &&
         initializeSettingsPopout(
             donutState.modControls.popout,
             donutState.modControls.tooltip,
-            animationDuration,
+            resources.animationDuration,
```
Comment (krasen86) [inline src/renderer.js:169]: Maybe if the animation duration is refactored to resources we don't need to pass it as a parameter, instead, just import it in the popout file?

File: src/renderer.js [hunk 2]
```diff
@@ -146,90 +123,59 @@ export async function render(donutState, modProperty) {
...
             modProperty
         );
 
     marker.drawRectangularSelection(donutState);
-    applyHoverEffect(pie, donutState, animationDuration);
-    addLabels(arc, pie, donutState, modProperty, animationDuration);
-    drawOuterLinesForNegativeValues(pie, donutState, animationDuration, padding, svg);
+    applyHoverEffect(pie, donutState, resources.animationDuration);
```
Comment (krasen86) [inline src/renderer.js:174]: Maybe here as well? We can just remove the animationDuration and add an import? Also for the labels?

File: src/labels.js [hunk 3]
```diff
@@ -40,11 +41,19 @@ export function addLabels(arc, pie, donutState, modProperty, animationDuration)
...
-        .text(modProperty.labelsVisible.value === "none" ? "" : getLabelText())
+        .text(modProperty.labelsVisible.value === resources.popoutLabelsVisibleNoneValue ? "" : getLabelText())
         .attr("transform", calculateLabelPosition)
-        .style("text-anchor", (d) =>
-            modProperty.labelsPosition.value() === "inside" ? "middle" : midAngle(d) < Math.PI ? "start" : "end"
-        )
+        .style("text-anchor", (d) => {
+            if (modProperty.labelsPosition.value() === resources.popoutLabelsPositionInsideValue) {
+                return "middle";
+            } else if (modProperty.circleType.value() === resources.popoutCircleTypeSemiValue) {
```
[the remark this answers is not in the document]
Comment (chrytsa) [reply src/labels.js:48]: Nice catch testing it with longer strings on labels! You 're right, middle does not work for all the cases!

[the remark this answers is not in the document]
Comment (chrytsa) [reply src/labels.js]: Just updated the calculation of the value depending on the corresponding threshold, defined by whether it being whole-circle or semi-circle.

**Review by clementinejensen — COMMENTED** [review]
I am wondering if the text is slightly more to the right than being in the center? I might have been staring at it for too long, but when I come back to look at it, it looks a bit off-center.

middle-value-offset

**Comment by chrytsa** [comment]
> I am wondering if the text is slightly more to the right than being in the center? I might have been staring at it for too long, but when I come back to look at it, it looks a bit off-center.
> 
> middle-value-offset
> 
> Other than this very small thing that I do not think anyone will notice, it looks great!

Which text, labels-related or center-text related? To be honest, I can't spot the difference. Also, there is no change between the older versions as far as the alignment on the X-axis. The correlated transformations introduced were only on the Y axis. Also, what are the different branches you are comparing with, because this doesn't exist on `dev` obviously?! :sweat_smile:

**Comment by clementinejensen** [comment]
> Which text, labels-related or center-text related? To be honest, I can't spot the difference. Also, there is no change between the older versions as far as the alignment on the X-axis. The correlated transformations introduced were only on the Y axis. Also, what are the different branches you are comparing with, because this doesn't exist on `dev` obviously?! 😅

Oh, I probably should have been a lot more specific! My bad! I am thinking about the center text, and I think I might just have stared at it for too long. It definitively does not exist on dev, it was just a quick change I did in my paint to see if I saw a difference or not, but I think, now that I am looking at it, that it looks good in your version! Ignore this 😆

**Comment by chrytsa** [comment]
> > Which text, labels-related or center-text related? To be honest, I can't spot the difference. Also, there is no change between the older versions as far as the alignment on the X-axis. The correlated transformations introduced were only on the Y axis. Also, what are the different branches you are comparing with, because this doesn't exist on `dev` obviously?! sweat_smile
> 
> Oh, I probably should have been a lot more specific! My bad! I am thinking about the center text, and I think I might just have stared at it for too long. It definitively does not exist on dev, it was just a quick change I did in my paint to see if I saw a difference or not, but I think, now that I am looking at it, that it looks good in your version! Ignore this laughing

Alright then! I thought that you might have compared it with the other branch that is up that has the older implementation. However, those two wouldn't match either, so I was wondering about the comparison between the `old` and the `new` version in your image! :stuck_out_tongue:

File: src/labels.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (krasen86) [reply src/labels.js]: NIce, it looks much better

**Review by krasen86 — APPROVED** [review]
Looks great, great work!

**Pull request merged**